pkg/services/correlations/correlations.go

```diff
@@ -0,0 +1,74 @@
+package correlations
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/api/routing"
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/events"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+)
+
+func ProvideService(sqlStore *sqlstore.SQLStore, routeRegister routing.RouteRegister, ds datasources.DataSourceService, ac accesscontrol.AccessControl, bus bus.Bus) *CorrelationsService {
+	s := &CorrelationsService{
+		SQLStore:          sqlStore,
+		RouteRegister:     routeRegister,
+		log:               log.New("correlations"),
+		DataSourceService: ds,
+		AccessControl:     ac,
+	}
+
+	s.registerAPIEndpoints()
+
+	bus.AddEventListener(s.handleDatasourceDeletion)
+
+	return s
+}
+
+type Service interface {
+	CreateCorrelation(ctx context.Context, cmd CreateCorrelationCommand) (Correlation, error)
+	DeleteCorrelationsBySourceUID(ctx context.Context, cmd DeleteCorrelationsBySourceUIDCommand) error
+	DeleteCorrelationsByTargetUID(ctx context.Context, cmd DeleteCorrelationsByTargetUIDCommand) error
+}
```

Comment:
If you can keep the types used here and this interface in this package and move everything else into a subdirectory you'll reduce the risk of having problems with circular dependencies messing things up a bit— it'll allow you to do things like define errors in this package and use them from services which are directly or indirectly dependencies of this service.

[contribute/architecture/backend/package-hierarchy.md](https://github.com/grafana/grafana/blob/53b8e528fc380a2f2bd32839208473c70956c142/contribute/architecture/backend/package-hierarchy.md#principles-of-how-to-structure-a-service-in-grafana) explains that a bit more.

Reply:
if that's ok with you, i'd rather do it in a follow up PR after i get other PRs depending on (and extending) this one in.